clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2512,13 +2548,38 @@ namespace {
       bool PreviousIncludeNullTerminator = IncludeNullTerminator;
       IncludeNullTerminator = IncludeNullTerm;
 
-      // Infer the lvalue or rvalue bounds of the subexpression.
+      // Infer the lvalue or rvalue bounds of the subexpression e1,
+      // setting State to contain the results for e1.
       BoundsExpr *SubExprTargetBounds, *SubExprLValueBounds, *SubExprBounds;
       InferBounds(SubExpr, CSS, SubExprTargetBounds,
                   SubExprLValueBounds, SubExprBounds, State);
 
       IncludeNullTerminator = PreviousIncludeNullTerminator;
 
+      // Update the set State.G of expressions that produce the
+      // same value as e.
+      if (CK == CastKind::CK_ArrayToPointerDecay)
+        // State.G = { e } for lvalues with array type.
+        State.G = { E };
```

Comment:
This doesn't account for adding the use of a temporary to State.G. For example, for the expression `ArrayToPointerDecay(Temp("abc"))`, State.G will only contain `ArrayToPointerDecay(Temp("abc"))`. It will not include `BoundsValue(Temp("abc"))`. The spec doesn't mention adding the use of any temporaries for ArrayToPointerDecay or LValueToRValue casts (or any lvalues). Should uses of temporaries be added to G in these cases?

Reply:
No, you don't need to do this.   There are two reasons.  First, when we lexicographically compare expression, we'll treat ArrayToPointerDecay(Temp("abc")) as equivalent to ArrayToPointerDeclary(BoundsValue(Temp("abc")).  See CompareExpr around line 222 in AST\CanonBounds.cpp.  Second, I don't believe you'll ever end up with BoundsValue by expression.  It'll still need to be wrapped in the ArrayToPointerDecay operator.